apps/heft/src/plugins/TypeScriptPlugin/TypeScriptBuilder.ts

```diff
@@ -794,15 +805,135 @@ export class TypeScriptBuilder extends SubprocessRunnerBase<ITypeScriptBuilderCo
+
+      function runWithSaveRestoreOptions<T>(fn: () => T): T {
+        const original: TTypescript.CompilerOptions = {};
+        Object.assign(original, compilerOptions);
+        try {
+          return fn();
+        } finally {
+          Object.assign(compilerOptions, original);
+        }
+      }
+
+      let defaultModuleKindResult: TTypescript.EmitResult;
+      const diagnostics: TTypescript.Diagnostic[] = [];
+      let emitSkipped: boolean = true;
+      for (const moduleKindToEmit of this._moduleKindsToEmit) {
+        const flavorResult: TTypescript.EmitResult = runWithSaveRestoreOptions(() => {
+          if (!moduleKindToEmit.isPrimary) {
+            compilerOptions.module = moduleKindToEmit.moduleKind;
+            compilerOptions.outDir = moduleKindToEmit.outFolderPath;
+            compilerOptions.declaration = false;
+            compilerOptions.declarationMap = false;
+          }
+
+          args[1] = wrapWriteFile(moduleKindToEmit.jsExtensionOverride);
+          return originalEmit(...args);
+        });
+
+        emitSkipped = emitSkipped && flavorResult.emitSkipped;
+        for (const diagnostic of flavorResult.diagnostics) {
+          diagnostics.push(diagnostic);
+        }
+
+        if (moduleKindToEmit.moduleKind === defaultModuleKind) {
+          defaultModuleKindResult = flavorResult;
+        }
+      }
+
+      const mergedDiagnostics: readonly TTypescript.Diagnostic[] =
+        ts.sortAndDeduplicateDiagnostics(diagnostics);
+
+      return {
+        ...defaultModuleKindResult!,
+        diagnostics: mergedDiagnostics,
+        emitSkipped
+      };
+    };
+    return program;
+  }
+
+  private _defaultModuleKind(): TTypescript.ModuleKind {
+    let foundPrimary: boolean = false;
+    let defaultModuleKind!: TTypescript.ModuleKind;
+
+    for (const moduleKindToEmit of this._moduleKindsToEmit) {
+      if (moduleKindToEmit.isPrimary) {
+        if (foundPrimary) {
+          throw new Error('Multiple primary module emit kinds encountered.');
+        } else {
+          foundPrimary = true;
+        }
+
+        defaultModuleKind = moduleKindToEmit.moduleKind;
+      }
+    }
+
+    return defaultModuleKind;
+  }
+
   private _emit(
     ts: ExtendedTypeScript,
-    tsconfig: TTypescript.ParsedCommandLine,
     genericProgram: TTypescript.BuilderProgram | TTypescript.Program
   ): IExtendedEmitResult {
     const filesToWrite: IFileToWrite[] = [];
 
-    const changedFiles: Set<IExtendedSourceFile> = new Set<IExtendedSourceFile>();
-    EmitFilesPatch.install(ts, tsconfig, this._moduleKindsToEmit, changedFiles);
+    const sourceFileMayBeEmittedHost: ISourceFileMayBeEmittedHost = {
+      getCompilerOptions: genericProgram.getCompilerOptions,
+      getResolvedProjectReferenceToRedirect: () => undefined,
+      isSourceFileFromExternalLibrary: () => false,
+      isSourceOfProjectReferenceRedirect: () => false
+    };
+
+    const changedFiles: Set<IExtendedSourceFile> = new Set<IExtendedSourceFile>(
```

Comment:
This whole thing is to just preserve the behavior of running ESLint only on files which had emit. I'm not actually sure that this is desirable though, so I would prefer to remove all of this and not depend on this internal API.

Reply:
We don't actually need to hack this, since we can still aggregate the information from the first argument to `emit`; in a builder program it is called on each affected source file separately.